homeassistant/components/mqtt/__init__.py

```diff
@@ -418,35 +435,50 @@ def async_publish_service(call):
     return True
 
 
-Subscription = namedtuple('Subscription',
-                          ['topic', 'callback', 'qos', 'encoding'])
-Subscription.__new__.__defaults__ = (0, 'utf-8')
+@attr.s(slots=True, frozen=True)
+class Subscription(object):
+    """Class to hold data about an active subscription."""
 
-Message = namedtuple('Message', ['topic', 'payload', 'qos', 'retain'])
-Message.__new__.__defaults__ = (0, False)
+    topic = attr.ib(type=str)
+    callback = attr.ib(type=MessageCallbackType)
+    qos = attr.ib(type=int, default=0)
+    encoding = attr.ib(type=str, default='utf-8')
+
+
+@attr.s(slots=True, frozen=True)
+class Message(object):
+    """MQTT Message."""
+
+    topic = attr.ib(type=str)
+    payload = attr.ib(type=PublishPayloadType)
+    qos = attr.ib(type=int, default=0)
+    retain = attr.ib(type=bool, default=False)
 
 
 class MQTT(object):
     """Home Assistant MQTT client."""
 
-    def __init__(self, hass, broker, port, client_id, keepalive, username,
-                 password, certificate, client_key, client_cert,
-                 tls_insecure, protocol, will_message: Optional[Message],
-                 birth_message: Optional[Message], tls_version):
+    def __init__(self, hass: HomeAssistantType, broker: str, port: int,
+                 client_id: Optional[str], keepalive: Optional[int],
+                 username: Optional[str], password: Optional[str],
+                 certificate: Optional[str], client_key: Optional[str],
+                 client_cert: Optional[str], tls_insecure: Optional[bool],
+                 protocol: Optional[str], will_message: Optional[Message],
+                 birth_message: Optional[Message], tls_version) -> None:
         """Initialize Home Assistant MQTT client."""
         import paho.mqtt.client as mqtt
 
         self.hass = hass
```

Comment:
Should we type class variables like this explicitly or can `mypy` find this out by itself from the argument typing? I couldn't find anything on this...